File: 2386-find-the-k-sum-of-an-array/2386-find-the-k-sum-of-an-array.cpp
class Solution {
public:
    long long kSum(vector<int>& nums, int k) {
        long long total = 0;
        for(int &i: nums) {
            if(i >= 0) total += i;
            i = abs(i);
        }
        int n = nums.size();
        sort(nums.begin(), nums.end());
        priority_queue<pair<long long,int>> pq;
        pq.push({-nums[0], 0});
        k -= 1;
        long long last = 0;
        while(k >= 1) {
            auto cur = pq.top();
            pq.pop();
            last = -cur.first;
            k--;
            int ind = cur.second;
            if(ind + 1 < n){
                pq.push({-(-cur.first + nums[ind + 1]), ind + 1});
                pq.push({-(nums[ind + 1] + last - nums[ind]),ind+1});
            }
        }
        
        return total - last;
    }
};

File: 2386-find-the-k-sum-of-an-array/2386-find-the-k-sum-of-an-array.py
class Solution:
    def kSum(self, nums: List[int], k: int) -> int:
        
        total = 0
        n = len(nums)
        for i in range(n):
            total += max(0, nums[i]) 
            nums[i] = abs(nums[i])
            
        nums.sort()
        heap = [(nums[0], 0)] 
        k -= 1
        
        last = 0
        while k >= 1:
            sum, ind = heapq.heappop(heap) 
            last = sum
            
            k-=1
            if ind + 1 < n: 
                heapq.heappush(heap, (last + nums[ind + 1], ind + 1))
                heapq.heappush(heap, (last - nums[ind] + nums[ind + 1], ind + 1))
                
            
        return total - last